File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.js
function solution(jobs) {
    var answer = 0;
    const list = new Array()
    let time = 0
    let i =0;
    
    jobs.sort((a, b) => a[0] - b[0]);
    
    while (i < jobs.length || list.length > 0) {
        while (i < jobs.length && time >= jobs[i][0]) {
            list.push(jobs[i])
            i += 1
        }
        if (list.length > 0){
            list.sort((a, b) => a[1] - b[1]);
            job = list.shift()
            console.log(job,job[1] + time - job[0])
            answer += job[1] + time - job[0]
            time += job[1]
        }
        else if (i < jobs.length) {
            time = jobs[i][0]
        }
    }
    
    return Math.floor(answer/jobs.length);
}

File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.py
import heapq

def solution(jobs):
    jobs_with_idx = [(arrival, duration, idx) for idx, (arrival, duration) in enumerate(jobs)]
    jobs_with_idx.sort(key=lambda x: x[0])

    n = len(jobs_with_idx)
    time = 0          
    i = 0               
    count = 0            
    total_turnaround = 0  
    heap = []         

    while count < n:
        while i < n and jobs_with_idx[i][0] <= time:
            arrival, duration, idx = jobs_with_idx[i]
            heapq.heappush(heap, (duration, arrival, idx))
            i += 1

        if not heap:
            time = jobs_with_idx[i][0]
            continue

        duration, arrival, idx = heapq.heappop(heap)
        time += duration
        total_turnaround += (time - arrival)
        count += 1

    return total_turnaround // n